ZEP000-wf-stream-processing.md

```diff
@@ -0,0 +1,1421 @@
+
+Notes:
+* an error can also be thrown by an error end event, that can be processed
+  through `EndEvent:Activate_Element` which results in
+  `EndEvent:Element_Activating`, `EndEvent:Element_Activated` and
+  `StartEvent:Activate_Element`
+
+On processing:
+* the `Job:Throw_Error` command marks the error on the job, makes the job non-activatable
+* the `ServiceTask:Terminate_Element` command transitions the service task element to terminated
+
+On replay:
+* the `Job:Error_Thrown` event marks the error on the job, makes the job non-activatable
+* the `ServiceTask:Element_Terminating` event transitions the service task element to terminating
+* the `ServiceTask:Element_Terminated` event transitions the service task element to terminated
+* processing can start again at `StartEvent:Activate_Element`
+
+Differences to current engine:
+* currently, the `Job:Error_Thrown` record is processed resulting in an event
+  trigger and the `Event_Occurred` record on the start event
+
+Considered alternatives:
+* nothing to mention
+
+### Message Start Event
+
+Command to Process | Written Events and Follow-up Commands
+---|---
+MessageStartEventSubscription:Open | MessageStartEventSubscription:Opened
+Message:Publish | Message:Published, MessageStartEventSubscription:Correlated, [Message:Deleted], Process:Activate_Element
+
+Notes:
+* the above table assumes the message is correlated to a message start event
+  subscription, if it isn't then it might be correlated to another message
+  event, or be ignored and deleted without correlation
+* `WorkflowInstanceSubscription` is not necessary like in message intermediate
+  catch event, because a new workflow instance can be created on any partition
+* the `Message:Deleted` event is only written when the message is not allowed to
+  live after it was published (i.e. `ttl = 0`). However, it will also be written
+  by a separate process to checks for expired messages. Therefore, when writing
+  a `Message:Deleted` event, the message should be removed from storage
+* the replay of `Message:Deleted` event should be idempotent. In this example
+  this event is only written if the message was not allowed to live, i.e. on
+  replay there is no message to delete. If the message was stored for some time
+  to live, it can simply be deleted on replay of the event. Deleting the message
+  on replay should thus be idempotent.
+
+On processing:
+* the `MessageStartEventSubscription:Open` command opens a message start event
+  subscription
+* the `Message:Publish` command stores the message when allowed to live (ttl>0),
+  correlates the message to the start event subscription, triggers the start
+  event and creates a new workflow instance, locks the workflow for this
+  correlation key
+
+On replay:
+* the `MessageStartEventSubscription:Opened` event opens a message start event
+  subscription
+* the `Message:Published` event stores the message when allowed to live (ttl>0)
```

Comment:
I don't think it's a bug because the logic is a simple if statement on a value in the record.

However, it's definitely indicative of business logic in the event replay. We've noticed that there are some other places where just applying state from an event means that you may need to calculate something which kinda goes against this proposal of not having business logic in replay. 

We can resolve it using your earlier suggestion: don't write `Message:Published` for records with `ttl = 0` but just write `Message:Ignored` or `Message:Correlated` if it was correlated immediately. Later on, published messages still need to be deleted.

For now, I think we can do 3 things: 
 - accept a simple if statement in replay of a specific event
 - revisit these records now and make some breaking changes
 - revisit these records later and make breaking changes later

Considering what this proposal is trying to achieve my preference would be to revisit now and make a breaking change now. It seems to be part of making this proposal work. @pihme WDYT?

Reply:
I prefer the first option "accept a simple if statement in replay of a specific event".

It keeps the lifecycle simple and consistent. Modelling this case with special intents adds more complexity to the consumers of the stream. 

Let's add it under open questions :writing_hand: